### ISSUE TITLE
[CS] func_vehicle knife hit sound bug

### ISSUE THREAD

**Issue by eat1k** [ValveSoftware/halflife#2178]
The bug consists in when you hit an entity func_vehicle with knife it emits a sound like you hit a player.

Solution: emit the sound "weapons/knife_hitwall1.wav".

Fix in AMX Mod X:

```
#include <amxmodx>
#include <fakemeta>
#include <hamsandwich>

new g_iMaxPlayers, g_iFhEmitSound
#define IsPlayer(%0) (1 <= %0 <= g_iMaxPlayers)

public plugin_init()
{
	register_plugin("Vehicle Knife Sound Fix", "1.0", "KayDee")
	register_cvar("@VehicleKnifeSoundFix", "1.0", FCVAR_SERVER|FCVAR_SPONLY|FCVAR_UNLOGGED)
	
	if(engfunc(EngFunc_FindEntityByString, -1, "classname", "func_vehicle"))
	{
		RegisterHam(Ham_TraceAttack, "func_vehicle", "Vehicle_TraceAttack", 1)
	
		g_iMaxPlayers = get_maxplayers()
	}
}

public Vehicle_TraceAttack(iVictim, iAttacker)
{
	if(IsPlayer(iAttacker) && get_user_weapon(iAttacker) == CSW_KNIFE)
	{
		g_iFhEmitSound = register_forward(FM_EmitSound, "Fm_EmitSound", 0)
	}
}

public Fm_EmitSound(iEnt, iChannel)
{
	emit_sound(iEnt, iChannel, "weapons/knife_hitwall1.wav", 1.0, ATTN_NORM, 0, PITCH_NORM)			
	unregister_forward(FM_EmitSound, g_iFhEmitSound, 0)	
	return FMRES_SUPERCEDE
}
```
Source: https://forums.alliedmods.net/showthread.php?p=2530647

**Comment by ArmynC** [comment]
Tested it on cart_attack and the bug is present.

**Comment by SamVanheer** [comment]
Looks like this happens because `func_vehicle` reports itself as class 14, while the Knife only does wall hit sounds if it's class 0 (none) or 1 (machine). This code was inherited from the original game: https://github.com/ValveSoftware/halflife/blob/5d761709a31ce1e71488f2668321de05f791b405/dlls/crowbar.cpp#L251

Adding a check for class 14 should fix it. No other entities are that class as far as i can tell so it shouldn't break any other entities, but check just to be sure.

Note that this has been the case since release, and it's well known so i'm not so sure if this should be fixed.

Maybe add a way to restore current behavior if you fix this.

**Comment by ArmynC** [comment]
> Maybe add a way to restore current behavior if you fix this.

I don't see this as keepable...

**Comment by SamVanheer** [comment]
Some mappers may have made func_vehicle entities that are supposed to be organic, in which case changing the sound will deviate from intended behavior.
Allowing both sound types to be played depending on the mapper's intention would be ideal.

**Comment by ArmynC** [comment]
Ah I see, like a horse, dragon. So, no one can check which one is inorganic nor organic.. The majority of them are cars, trains etc. It needs to set it as normal objects with the option to have that sound if it is desired. This will affect that small number of maps which will not be updated.. still, it is better to fix most of them.

PS: These are pushable ones but are some objects that aren't pushable, just 'callable' with a pre-set route (see helicopters), do we need that sound here too?

**Comment by SamVanheer** [comment]
Only func_vehicle returns class 14, the default is to return 0 (none) so those should all work fine.

func_guntarget has class 1 (machine)
Players have class 2
Hostages have class 3
monster_hevsuit_dead has class 4

**Comment by ArmynC** [comment]
Yeah I know, there are just the pushable ones, where on pre-seted objects (auto helicopters) there isn't this sound but we should have here too a value to set that sound (if we got a dragon for example). More like a feature...
